File: 2023/day05/php/solution.php
<?php

declare(strict_types=1);

/**
 * Advent of Code 2023 - Day 5: If You Give A Seed A Fertilizer
 *
 * This solution handles seed-to-location mapping through a series of
 * transformation maps. Part 1 processes individual seeds, while Part 2
 * efficiently handles billions of seeds by operating on ranges.
 */

/**
 * Parse the input into seeds array and list of transformation maps.
 *
 * Each map contains ranges with destination start, source start, and length.
 *
 * @param string $text Raw input text
 * @return array{0: int[], 1: array<array<array{dst_start: int, src_start: int, length: int}>>}
 */
function parseInput(string $text): array
{
    $sections = explode("\n\n", trim($text));

    // Parse seeds from first section: "seeds: 79 14 55 13"
    $seedPart = explode(': ', $sections[0])[1];
    $seeds = array_map(intval(...), preg_split('/\s+/', trim($seedPart)));

    // Parse each mapping section (seed-to-soil, soil-to-fertilizer, etc.)
    $maps = array_map(
        function (string $section): array {
            $lines = explode("\n", trim($section));
            // Skip the header line (e.g., "seed-to-soil map:")
            return array_map(
                function (string $line): array {
                    $parts = array_map(intval(...), preg_split('/\s+/', trim($line)));
                    return [
                        'dst_start' => $parts[0],
                        'src_start' => $parts[1],
                        'length' => $parts[2],
                    ];
                },
                array_slice($lines, 1)
            );
        },
        array_slice($sections, 1)
    );

    return [$seeds, $maps];
}

/**
 * Apply a single transformation map to a value.
 *
 * If the value falls within any source range, it gets mapped to the
 * corresponding destination range. Otherwise, it passes through unchanged.
 *
 * @param int $value The value to transform
 * @param array<array{dst_start: int, src_start: int, length: int}> $ranges The mapping ranges
 * @return int The transformed value
 */
function applyMap(int $value, array $ranges): int
{
    foreach ($ranges as $range) {
        $srcStart = $range['src_start'];
        $srcEnd = $srcStart + $range['length'];

        if ($value >= $srcStart && $value < $srcEnd) {
            return $range['dst_start'] + ($value - $srcStart);
        }
    }

    // No matching range found; value maps to itself
    return $value;
}

/**
 * Convert a seed number to a location number through all transformation maps.
 *
 * @param int $seed The initial seed number
 * @param array<array<array{dst_start: int, src_start: int, length: int}>> $maps All transformation maps
 * @return int The final location number
 */
function seedToLocation(int $seed, array $maps): int
{
    return array_reduce(
        $maps,
        fn(int $value, array $mapRanges): int => applyMap($value, $mapRanges),
        $seed
    );
}

/**
 * Part 1: Find the lowest location number for any initial seed.
 *
 * @param int[] $seeds List of seed numbers
 * @param array<array<array{dst_start: int, src_start: int, length: int}>> $maps All transformation maps
 * @return int The minimum location number
 */
function part1(array $seeds, array $maps): int
{
    $locations = array_map(
        fn(int $seed): int => seedToLocation($seed, $maps),
        $seeds
    );

    return min($locations);
}

/**
 * Apply a transformation map to a list of ranges, producing new ranges.
 *
 * Range-Splitting Algorithm:
 * -------------------------
 * For each input range [start, end), we check it against each mapping range.
 * A mapping range transforms values from [src_start, src_end) to [dst_start, dst_end).
 *
 * When an input range overlaps with a mapping range, we split it into up to 3 parts:
 *
 *   Input range:     |-----------|
 *   Mapping range:        |-----|
 *   Result:          |---|xxxxx|--|
 *                     ^    ^     ^
 *                     |    |     |
 *                     |    |     +-- Part after: [max(start, src_end), end) - stays unmapped
 *                     |    +-------- Overlap: [max(start, src_start), min(end, src_end)) - gets transformed
 *                     +------------- Part before: [start, min(end, src_start)) - stays unmapped
 *
 * The "before" and "after" parts continue to be checked against remaining mapping ranges.
 * The overlapping part is transformed and added directly to results.
 * Any parts that don't match any mapping range pass through unchanged (identity mapping).
 *
 * @param array<array{0: int, 1: int}> $inputRanges Ranges as [start, end) pairs
 * @param array<array{dst_start: int, src_start: int, length: int}> $mapRanges The mapping ranges
 * @return array<array{0: int, 1: int}> Transformed ranges
 */
function applyMapToRanges(array $inputRanges, array $mapRanges): array
{
    $result = [];

    foreach ($inputRanges as [$start, $end]) {
        // Track parts of this range that haven't been mapped yet
        $remaining = [[$start, $end]];

        // Try each mapping range against the unmapped parts
        foreach ($mapRanges as $range) {
            $dstStart = $range['dst_start'];
            $srcStart = $range['src_start'];
            $srcEnd = $srcStart + $range['length'];
            $offset = $dstStart - $srcStart;

            $newRemaining = [];

            foreach ($remaining as [$rStart, $rEnd]) {
                // Part before the mapping range (stays unmapped, try other ranges)
                if ($rStart < $srcStart) {
                    $newRemaining[] = [$rStart, min($rEnd, $srcStart)];
                }

                // Overlapping part (gets transformed, goes directly to result)
                $overlapStart = max($rStart, $srcStart);
                $overlapEnd = min($rEnd, $srcEnd);
                if ($overlapStart < $overlapEnd) {
                    $result[] = [$overlapStart + $offset, $overlapEnd + $offset];
                }

                // Part after the mapping range (stays unmapped, try other ranges)
                if ($rEnd > $srcEnd) {
                    $newRemaining[] = [max($rStart, $srcEnd), $rEnd];
                }
            }

            $remaining = $newRemaining;
        }

        // Anything still remaining maps to itself (identity)
        array_push($result, ...$remaining);
    }

    return $result;
}

/**
 * Part 2: Find the lowest location for seed ranges.
 *
 * Instead of individual seeds, the input is interpreted as pairs of
 * (start, length) defining ranges of seeds. This could represent billions
 * of seeds, so we operate on ranges rather than individual values.
 *
 * @param int[] $seeds Seed range specifications as [start1, len1, start2, len2, ...]
 * @param array<array<array{dst_start: int, src_start: int, length: int}>> $maps All transformation maps
 * @return int The minimum location number
 */
function part2(array $seeds, array $maps): int
{
    // Convert seed pairs to ranges: [start, start + length)
    $ranges = [];
    for ($i = 0, $len = count($seeds); $i < $len; $i += 2) {
        $ranges[] = [$seeds[$i], $seeds[$i] + $seeds[$i + 1]];
    }

    // Transform ranges through each map
    $ranges = array_reduce(
        $maps,
        fn(array $ranges, array $mapRanges): array => applyMapToRanges($ranges, $mapRanges),
        $ranges
    );

    // Find minimum start value across all resulting ranges
    return min(array_column($ranges, 0));
}

/**
 * Main entry point.
 */
function main(): void
{
    $inputPath = __DIR__ . '/../input.txt';
    $text = file_get_contents($inputPath);

    [$seeds, $maps] = parseInput($text);

    echo 'Part 1: ' . part1($seeds, $maps) . PHP_EOL;
    echo 'Part 2: ' . part2($seeds, $maps) . PHP_EOL;
}

main();

File: 2023/day05/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

/**
 * Advent of Code 2023 - Day 5: If You Give A Seed A Fertilizer
 *
 * Transforms seed numbers through a series of category mappings to find
 * the minimum location number.
 */
public class Solution {

    /**
     * Represents a mapping range that transforms source values to destination values.
     * Values in [srcStart, srcStart + length) map to [dstStart, dstStart + length).
     *
     * @param dstStart the start of the destination range
     * @param srcStart the start of the source range
     * @param length   the length of both ranges
     */
    record MapRange(long dstStart, long srcStart, long length) {

        /** Returns the exclusive end of the source range. */
        long srcEnd() {
            return srcStart + length;
        }

        /** Returns the offset to add when mapping source to destination. */
        long offset() {
            return dstStart - srcStart;
        }

        /** Checks if the given value falls within this mapping's source range. */
        boolean contains(long value) {
            return value >= srcStart && value < srcEnd();
        }
    }

    /**
     * Represents a half-open range [start, end) of values.
     *
     * @param start inclusive start of the range
     * @param end   exclusive end of the range
     */
    record Range(long start, long end) {

        /** Returns true if this range is non-empty. */
        boolean isValid() {
            return start < end;
        }
    }

    public static void main(String[] args) throws IOException {
        String text = Files.readString(Path.of("../input.txt"));
        String[] sections = text.strip().split("\n\n");

        long[] seeds = parseSeeds(sections[0]);
        List<List<MapRange>> maps = parseMaps(sections);

        System.out.println("Part 1: " + part1(seeds, maps));
        System.out.println("Part 2: " + part2(seeds, maps));
    }

    /**
     * Parses the seed line into an array of seed numbers.
     */
    private static long[] parseSeeds(String seedLine) {
        return Arrays.stream(seedLine.split(": ")[1].split("\\s+"))
                .mapToLong(Long::parseLong)
                .toArray();
    }

    /**
     * Parses all mapping sections into lists of MapRange objects.
     */
    private static List<List<MapRange>> parseMaps(String[] sections) {
        return IntStream.range(1, sections.length)
                .mapToObj(i -> parseMapSection(sections[i]))
                .toList();
    }

    /**
     * Parses a single map section into a list of MapRange objects.
     */
    private static List<MapRange> parseMapSection(String section) {
        return Arrays.stream(section.strip().split("\n"))
                .skip(1) // Skip the header line
                .map(Solution::parseMapRange)
                .toList();
    }

    /**
     * Parses a single line into a MapRange.
     */
    private static MapRange parseMapRange(String line) {
        String[] parts = line.split("\\s+");
        return new MapRange(
                Long.parseLong(parts[0]),
                Long.parseLong(parts[1]),
                Long.parseLong(parts[2])
        );
    }

    /**
     * Applies a single mapping to a value, returning the mapped result.
     * If no mapping applies, returns the original value (identity mapping).
     */
    private static long applyMap(long value, List<MapRange> ranges) {
        return ranges.stream()
                .filter(range -> range.contains(value))
                .findFirst()
                .map(range -> range.dstStart() + (value - range.srcStart()))
                .orElse(value);
    }

    /**
     * Transforms a seed through all category mappings to get its final location.
     */
    private static long seedToLocation(long seed, List<List<MapRange>> maps) {
        long value = seed;
        for (List<MapRange> mapRanges : maps) {
            value = applyMap(value, mapRanges);
        }
        return value;
    }

    /**
     * Part 1: Find the minimum location for individual seed values.
     */
    private static long part1(long[] seeds, List<List<MapRange>> maps) {
        return Arrays.stream(seeds)
                .map(seed -> seedToLocation(seed, maps))
                .min()
                .orElseThrow();
    }

    /**
     * Applies a mapping to a list of ranges, producing transformed ranges.
     * Handles splitting ranges at mapping boundaries and applying offsets.
     */
    private static List<Range> applyMapToRanges(List<Range> inputRanges, List<MapRange> mapRanges) {
        List<Range> result = new ArrayList<>();

        for (Range input : inputRanges) {
            List<Range> remaining = new ArrayList<>(List.of(input));

            for (MapRange mr : mapRanges) {
                List<Range> newRemaining = new ArrayList<>();

                for (Range r : remaining) {
                    // Part before the map range (unmapped, stays for next iteration)
                    Range before = new Range(r.start(), Math.min(r.end(), mr.srcStart()));
                    if (before.isValid()) {
                        newRemaining.add(before);
                    }

                    // Part within the map range (mapped to result)
                    long overlapStart = Math.max(r.start(), mr.srcStart());
                    long overlapEnd = Math.min(r.end(), mr.srcEnd());
                    if (overlapStart < overlapEnd) {
                        result.add(new Range(
                                overlapStart + mr.offset(),
                                overlapEnd + mr.offset()
                        ));
                    }

                    // Part after the map range (unmapped, stays for next iteration)
                    Range after = new Range(Math.max(r.start(), mr.srcEnd()), r.end());
                    if (after.isValid()) {
                        newRemaining.add(after);
                    }
                }

                remaining = newRemaining;
            }

            // Any remaining parts are unmapped (identity mapping)
            result.addAll(remaining);
        }

        return result;
    }

    /**
     * Part 2: Seeds are interpreted as pairs (start, length) defining ranges.
     * Find the minimum location across all seed ranges.
     */
    private static long part2(long[] seeds, List<List<MapRange>> maps) {
        // Convert seed pairs to ranges
        List<Range> ranges = IntStream.iterate(0, i -> i < seeds.length, i -> i + 2)
                .mapToObj(i -> new Range(seeds[i], seeds[i] + seeds[i + 1]))
                .collect(ArrayList::new, ArrayList::add, ArrayList::addAll);

        // Apply each map to the ranges
        for (List<MapRange> mapRanges : maps) {
            ranges = applyMapToRanges(ranges, mapRanges);
        }

        // Find minimum start of any range
        return ranges.stream()
                .mapToLong(Range::start)
                .min()
                .orElseThrow();
    }
}
